--- configure.ac.old	2009-08-22 20:01:53.000000000 +0200
+++ configure.ac	2009-08-22 20:02:32.000000000 +0200
@@ -220,14 +220,31 @@
 #
 # Find keysymdef.h
 #
+KEYSYMDEF=""
 AC_MSG_CHECKING([keysymdef.h])
-dir=`pkg-config --variable=includedir xproto`
-KEYSYMDEF="$dir/X11/keysymdef.h"
-if test -f "$KEYSYMDEF"; then
+for flag in $XPROTO_CFLAGS -I/usr/include; do
+        case "$KEYSYMDEF" in
+        "")
+                case "$flag" in
+                *-I*)
+                        dir=`echo "$flag" | sed 's/ *-I//'`
+                        file="$dir/X11/keysymdef.h"
+                        if test -f "$file"; then
+                                KEYSYMDEF="$file"
+                        fi
+                        ;;
+                esac
+                ;;
+        esac
+done
+case "$KEYSYMDEF" in
+"")
+        AC_MSG_ERROR([Cannot find keysymdef.h])
+        ;;
+*)
         AC_MSG_RESULT([$KEYSYMDEF])
-else
-	AC_MSG_ERROR([Cannot find keysymdef.h])
-fi
+        ;;
+esac
 AC_SUBST(KEYSYMDEF)
 
 AM_CONDITIONAL(UDC, test xfalse = xtrue)
