--- gcc-4.3.3/libstdc++-v3/include/Makefile.am.orig	2009-04-04 21:33:37.000000000 +0200
+++ gcc-4.3.3/libstdc++-v3/include/Makefile.am	2009-04-04 21:36:36.000000000 +0200
@@ -1084,7 +1084,7 @@
 # Build a precompiled TR1 include, stdtr1c++.h.gch/O2.gch
 ${pch2_output}: ${pch2_source} ${pch1_output}
 	-mkdir -p ${pch2_output_builddir}
-	$(CXX) $(PCHFLAGS) $(AM_CPPFLAGS) -O2 -g ${pch2_source} -o $@
+	$(CXX) $(PCHFLAGS) $(AM_CPPFLAGS) -nostdinc++ -O2 -g  ${pch2_source} -o $@
 
 # Build a precompiled extension include, extc++.h.gch/O2.gch
 ${pch3_output}: ${pch3_source} ${pch2_output}
--- gcc-4.3.3/libstdc++-v3/include/Makefile.in.orig	2009-04-04 21:36:09.000000000 +0200
+++ gcc-4.3.3/libstdc++-v3/include/Makefile.in	2009-04-04 21:36:25.000000000 +0200
@@ -1475,7 +1475,7 @@
 # Build a precompiled TR1 include, stdtr1c++.h.gch/O2.gch
 ${pch2_output}: ${pch2_source} ${pch1_output}
 	-mkdir -p ${pch2_output_builddir}
-	$(CXX) $(PCHFLAGS) $(AM_CPPFLAGS) -O2 -g ${pch2_source} -o $@
+	$(CXX) $(PCHFLAGS) $(AM_CPPFLAGS) -nostdinc++ -O2 -g ${pch2_source} -o $@
 
 # Build a precompiled extension include, extc++.h.gch/O2.gch
 ${pch3_output}: ${pch3_source} ${pch2_output}
