--- python/setup.py	2003-11-18 18:55:09.000000000 +0000
+++ python/setup.py.new	2009-05-07 21:18:09.000000000 +0000
@@ -3,8 +3,8 @@
 
 magic_module = Extension('magic',
     libraries = ['magic'],
-    library_dirs = ['./','../','../src','/usr/lib/'],
-    include_dirs = ['./','../','../src','/usr/include/'],
+    library_dirs = ['./','../','../src/.libs'],
+    include_dirs = ['./','../','../src'],
     sources = ['py_magic.c'])
 
 setup (name = 'Magic file extensions',
