From 2d55e4c88d05968ff52a3e4d1e5931c6383f8c34 Mon Sep 17 00:00:00 2001
From: Jacob Godserv <jacobgodserv@gmail.com>
Date: Mon, 15 Mar 2010 20:46:23 -0400
Subject: [PATCH 03/52] Added some debug code, but need to be sure it's safe to add

---
 pym/portage/hooks.py |    5 +++++
 1 files changed, 5 insertions(+), 0 deletions(-)

diff --git a/pym/portage/hooks.py b/pym/portage/hooks.py
index 7ffdc82..b6126fa 100644
--- a/pym/portage/hooks.py
+++ b/pym/portage/hooks.py
@@ -2,6 +2,11 @@
 # Distributed under the terms of the GNU General Public License v2
 # $Id$
 
+# TODO: following may be harmful, but helpful for debugging
+import os, sys
+import os.path as osp
+sys.path.insert(0, osp.dirname(osp.dirname(osp.abspath(__file__))))
+
 from portage.const import BASH_BINARY, HOOKS_PATH
 from portage import os
 from portage import check_config_instance
-- 
1.6.4.4

