From d9a829ea52838e9e3adf474df7b3f14b1587d1db Mon Sep 17 00:00:00 2001
From: Jacob Godserv <jacobgodserv@gmail.com>
Date: Sun, 21 Mar 2010 19:49:48 -0400
Subject: [PATCH 14/52] Style adjustments, to make the position of the pre-run hook clearer

---
 pym/_emerge/main.py |    5 ++++-
 1 files changed, 4 insertions(+), 1 deletions(-)

diff --git a/pym/_emerge/main.py b/pym/_emerge/main.py
index d61d91e..90d13fe 100644
--- a/pym/_emerge/main.py
+++ b/pym/_emerge/main.py
@@ -1181,9 +1181,12 @@ def emerge_main():
 	# Portage needs to ensure a sane umask for the files it creates.
 	os.umask(0o22)
 	settings, trees, mtimedb = load_emerge_config()
+	portdb = trees[settings["ROOT"]]["porttree"].dbapi
+
+	# Portage configured; let's let a hook set everything up before we do anything more
 	if "hooks" in settings['FEATURES']:
 		portage.hooks.HookDirectory(phase='pre-run', settings=settings, myopts=myopts, myaction=myaction, myfiles=myfiles).execute()
-	portdb = trees[settings["ROOT"]]["porttree"].dbapi
+
 	rval = profile_check(trees, myaction)
 	if rval != os.EX_OK:
 		return rval
-- 
1.6.4.4

