From 76bdf7463f8d72e4d4823d42071cba5234fd30d2 Mon Sep 17 00:00:00 2001
From: Jacob Godserv <jacobgodserv@gmail.com>
Date: Sun, 21 Mar 2010 20:03:58 -0400
Subject: [PATCH 16/52] Fixed a small bug where isolated-functions.sh didn't affect the hook's environment at all

---
 pym/portage/hooks.py |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/pym/portage/hooks.py b/pym/portage/hooks.py
index 4a5c6bd..4a44b6e 100644
--- a/pym/portage/hooks.py
+++ b/pym/portage/hooks.py
@@ -74,7 +74,7 @@ class HookFile (object):
 				for myfile in self.myfiles:
 					command.extend(['--file', myfile])
 			
-			command=[BASH_BINARY, '-c', 'source ' + PORTAGE_BIN_PATH + '/isolated-functions.sh && ' + ' '.join(command)]
+			command=[BASH_BINARY, '-c', 'source ' + PORTAGE_BIN_PATH + '/isolated-functions.sh && source ' + ' '.join(command)]
 			code = spawn(mycommand=command, env=self.settings.environ())
 			if code: # if failure
 				raise PortageException('!!! Hook %s failed with exit code %s' % (path, code))
-- 
1.6.4.4

