From 6dda2e189c2c3c08a251f8d7f7b6a2d0bce098cb Mon Sep 17 00:00:00 2001
From: Jacob Godserv <jacobgodserv@gmail.com>
Date: Thu, 27 May 2010 20:57:04 -0400
Subject: [PATCH 23/52] Added post-run hooks

---
 pym/_emerge/main.py |    2 ++
 1 files changed, 2 insertions(+), 0 deletions(-)

diff --git a/pym/_emerge/main.py b/pym/_emerge/main.py
index 6b14567..b90f212 100644
--- a/pym/_emerge/main.py
+++ b/pym/_emerge/main.py
@@ -29,6 +29,7 @@ import portage.elog
 import portage.dep
 portage.dep._dep_check_strict = True
 import portage.hooks
+import portage.process
 import portage.util
 import portage.locks
 import portage.exception
@@ -1184,6 +1185,7 @@ def emerge_main():
 	portdb = trees[settings["ROOT"]]["porttree"].dbapi
 
 	# Portage configured; let's let a hook set everything up before we do anything more
+	portage.process.atexit_register(portage.hooks.HookDirectory(phase='post-run', settings=settings, myopts=myopts, myaction=myaction, mytargets=myfiles).execute)
 	portage.hooks.HookDirectory(phase='pre-run', settings=settings, myopts=myopts, myaction=myaction, mytargets=myfiles).execute()
 
 	rval = profile_check(trees, myaction)
-- 
1.6.4.4

