From 70a675792a91a1bebe3707ad268dcaff0fa6e137 Mon Sep 17 00:00:00 2001
From: Jacob Godserv <jacobgodserv@gmail.com>
Date: Wed, 2 Jun 2010 16:31:43 -0400
Subject: [PATCH 31/52] Minor fixes to hooks commenting in main.py

---
 pym/_emerge/main.py |    3 ++-
 1 files changed, 2 insertions(+), 1 deletions(-)

diff --git a/pym/_emerge/main.py b/pym/_emerge/main.py
index b90f212..06c1d57 100644
--- a/pym/_emerge/main.py
+++ b/pym/_emerge/main.py
@@ -1184,8 +1184,9 @@ def emerge_main():
 	settings, trees, mtimedb = load_emerge_config()
 	portdb = trees[settings["ROOT"]]["porttree"].dbapi
 
-	# Portage configured; let's let a hook set everything up before we do anything more
+	# Have post-run hooks executed whenever portage quits
 	portage.process.atexit_register(portage.hooks.HookDirectory(phase='post-run', settings=settings, myopts=myopts, myaction=myaction, mytargets=myfiles).execute)
+	# Portage configured; let's let hooks run before we do anything more
 	portage.hooks.HookDirectory(phase='pre-run', settings=settings, myopts=myopts, myaction=myaction, mytargets=myfiles).execute()
 
 	rval = profile_check(trees, myaction)
-- 
1.6.4.4

