From 93500d1523d7e0a8998bd2c2b2cd3bd5e7e3746e Mon Sep 17 00:00:00 2001
From: Jacob Godserv <jacobgodserv@gmail.com>
Date: Wed, 2 Jun 2010 16:32:46 -0400
Subject: [PATCH 32/52] Removed debug code from hooks.py

---
 pym/portage/hooks.py |    9 ---------
 1 files changed, 0 insertions(+), 9 deletions(-)

diff --git a/pym/portage/hooks.py b/pym/portage/hooks.py
index 6dd55e6..ec4436d 100644
--- a/pym/portage/hooks.py
+++ b/pym/portage/hooks.py
@@ -2,11 +2,6 @@
 # Distributed under the terms of the GNU General Public License v2
 # $Id$
 
-# TODO: following may be harmful, but helpful for debugging
-#import os, sys
-#import os.path as osp
-#sys.path.insert(0, osp.dirname(osp.dirname(osp.abspath(__file__))))
-
 from portage.const import BASH_BINARY, HOOKS_PATH, PORTAGE_BIN_PATH
 from portage import os
 from portage import check_config_instance
@@ -87,7 +82,3 @@ class HookFile (object):
 		
 		else:
 			raise InvalidLocation('This hook path ought to be a file: ' + self.path)
-
-if __name__ == "__main__": # TODO: debug
-	from portage.package.ebuild.config import config
-	HookDirectory('run', config()).execute()
-- 
1.6.4.4

