From 79d2c2a1a4d6774fa36033b662c657957f416539 Mon Sep 17 00:00:00 2001
From: Jacob Godserv <jacobgodserv@gmail.com>
Date: Tue, 8 Jun 2010 17:42:03 -0400
Subject: [PATCH 52/52] Added mention of FEATURES="hooks" to documentation

---
 doc/config/hooks.docbook |    3 +++
 1 files changed, 3 insertions(+), 0 deletions(-)

diff --git a/doc/config/hooks.docbook b/doc/config/hooks.docbook
index 5017b67..5fc6000 100644
--- a/doc/config/hooks.docbook
+++ b/doc/config/hooks.docbook
@@ -4,6 +4,9 @@
 	<sect1 id='config-hooks-execution'>
 		<title id="config-hooks-execution.title">Hooks Execution</title>
 		
+		<para>Hooks are only executed if <quote>hooks</quote> is set in
+		FEATURES.</para>
+		
 		<para>
 		If a hook directory exists, the bash scripts within each one
 		wil either be executed before or after that particular phase, in
-- 
1.6.4.4

