--- setup.py	2009-11-21 22:12:27.000000000 +0000
+++ setup-sysroot.py	2009-11-21 22:19:11.000000000 +0000
@@ -337,9 +337,15 @@
 
     def detect_modules(self):
         global disable_ssl
+
+        # We must respect the user specified sysroot!
+        sysroot=os.getenv('SYSROOT')
+        if sysroot is None:
+                sysroot=''
+
         # Ensure that /usr/local is always used
-        add_dir_to_list(self.compiler.library_dirs, '/usr/local/lib')
-        add_dir_to_list(self.compiler.include_dirs, '/usr/local/include')
+        add_dir_to_list(self.compiler.library_dirs, sysroot+'/usr/local/lib')
+        add_dir_to_list(self.compiler.include_dirs, sysroot+'/usr/local/include')
 
         # Add paths specified in the environment variables LDFLAGS and
         # CPPFLAGS for header and library files.
@@ -613,11 +619,11 @@
             elif self.compiler.find_library_file(lib_dirs, 'curses'):
                 readline_libs.append('curses')
             elif self.compiler.find_library_file(lib_dirs +
-                                               ['/usr/@@GENTOO_LIBDIR@@/termcap'],
+                                               [sysroot+'/usr/@@GENTOO_LIBDIR@@/termcap'],
                                                'termcap'):
                 readline_libs.append('termcap')
             exts.append( Extension('readline', ['readline.c'],
-                                   library_dirs=['/usr/@@GENTOO_LIBDIR@@/termcap'],
+                                   library_dirs=[sysroot+'/usr/@@GENTOO_LIBDIR@@/termcap'],
                                    extra_link_args=readline_extra_link_args,
                                    libraries=readline_libs) )
         else:
@@ -642,20 +648,20 @@
                                depends = ['socketmodule.h']) )
         # Detect SSL support for the socket module (via _ssl)
         search_for_ssl_incs_in = [
-                              '/usr/local/ssl/include',
-                              '/usr/contrib/ssl/include/'
+                              sysroot+'/usr/local/ssl/include',
+                              sysroot+'/usr/contrib/ssl/include/'
                              ]
         ssl_incs = find_file('openssl/ssl.h', inc_dirs,
                              search_for_ssl_incs_in
                              )
         if ssl_incs is not None and not disable_ssl:
             krb5_h = find_file('krb5.h', inc_dirs,
-                               ['/usr/kerberos/include'])
+                               [sysroot+'/usr/kerberos/include'])
             if krb5_h:
                 ssl_incs += krb5_h
         ssl_libs = find_library_file(self.compiler, 'ssl',lib_dirs,
-                                     ['/usr/local/ssl/lib',
-                                      '/usr/contrib/ssl/lib/'
+                                     [sysroot+'/usr/local/ssl/lib',
+                                      sysroot+'/usr/contrib/ssl/lib/'
                                      ] )
 
         if (ssl_incs is not None and
