diff --git a/config/Makefile.am b/config/Makefile.am
index d79806d..86491e6 100644
--- a/config/Makefile.am
+++ b/config/Makefile.am
@@ -1,7 +1,11 @@
 MAINTAINERCLEANFILES = Makefile.in
 SUBDIRS = default standard minimalist scaleable netbook illume
 
+if CROSS_COMPILING
+EET_EET = @HOST_PROG_EET@
+else
 EET_EET = @eet_eet@
+endif
 
 EXTRA_DIST = profile.src
 
diff --git a/config/default/Makefile.am b/config/default/Makefile.am
index f28a8da..d9d0576 100644
--- a/config/default/Makefile.am
+++ b/config/default/Makefile.am
@@ -1,6 +1,10 @@
 MAINTAINERCLEANFILES = Makefile.in
 
+if CROSS_COMPILING
+EET_EET = @HOST_PROG_EET@
+else
 EET_EET = @eet_eet@
+endif
 
 EXTRA_DIST = \
         profile.desktop \
diff --git a/config/illume/Makefile.am b/config/illume/Makefile.am
index b6e9a61..d270e49 100644
--- a/config/illume/Makefile.am
+++ b/config/illume/Makefile.am
@@ -1,6 +1,10 @@
 MAINTAINERCLEANFILES = Makefile.in
 
+if CROSS_COMPILING
+EET_EET = @HOST_PROG_EET@
+else
 EET_EET = @eet_eet@
+endif
 
 EXTRA_DIST = \
         profile.desktop \
diff --git a/config/minimalist/Makefile.am b/config/minimalist/Makefile.am
index d453d5d..264d85f 100644
--- a/config/minimalist/Makefile.am
+++ b/config/minimalist/Makefile.am
@@ -1,6 +1,10 @@
 MAINTAINERCLEANFILES = Makefile.in
 
+if CROSS_COMPILING
+EET_EET = @HOST_PROG_EET@
+else
 EET_EET = @eet_eet@
+endif
 
 EXTRA_DIST = \
         profile.desktop \
diff --git a/config/netbook/Makefile.am b/config/netbook/Makefile.am
index 4ed2fe4..74a5dcd 100644
--- a/config/netbook/Makefile.am
+++ b/config/netbook/Makefile.am
@@ -1,6 +1,10 @@
 MAINTAINERCLEANFILES = Makefile.in
 
+if CROSS_COMPILING
+EET_EET = @HOST_PROG_EET@
+else
 EET_EET = @eet_eet@
+endif
 
 EXTRA_DIST = \
         profile.desktop \
diff --git a/config/scaleable/Makefile.am b/config/scaleable/Makefile.am
index 0952d2b..1e4736e 100644
--- a/config/scaleable/Makefile.am
+++ b/config/scaleable/Makefile.am
@@ -1,6 +1,10 @@
 MAINTAINERCLEANFILES = Makefile.in
 
+if CROSS_COMPILING
+EET_EET = @HOST_PROG_EET@
+else
 EET_EET = @eet_eet@
+endif
 
 EXTRA_DIST = \
         profile.desktop \
diff --git a/config/standard/Makefile.am b/config/standard/Makefile.am
index 0e6b0d0..820d73b 100644
--- a/config/standard/Makefile.am
+++ b/config/standard/Makefile.am
@@ -1,6 +1,6 @@
 MAINTAINERCLEANFILES = Makefile.in
 
-EET_EET = @eet_eet@
+EET_EET = @HOST_PROG_EET@
 
 EXTRA_DIST = \
         profile.desktop \
diff --git a/configure.ac b/configure.ac
index 31da04b..7eeb601 100644
--- a/configure.ac
+++ b/configure.ac
@@ -32,6 +32,11 @@ AC_SUBST(release)
 AC_CHECK_FUNCS(setenv)
 AC_CHECK_FUNCS(unsetenv)
 
+HOST_PROG_EDJE_CC=edje_cc
+HOST_PROG_EET=eet
+AC_SUBST([HOST_PROG_EDJE_CC])
+AC_SUBST([HOST_PROG_EET])
+
 AC_CHECK_HEADERS(fnmatch.h,, AC_MSG_ERROR([Cannot find fnmatch.h. Make sure your CFLAGS environment variable contains include lines for the location of this file]))
 
 AC_CHECK_FUNCS(fnmatch, res=yes, res=no)
@@ -334,6 +339,9 @@ define([CHECK_MODULE_CONF_THEME],
 ])
 
 AM_CONDITIONAL(HAVE_ALSA, false)
+
+AM_CONDITIONAL(CROSS_COMPILING, [ test $cross_compiling = yes ])
+
 define([CHECK_MODULE_MIXER],
 [
 if test "x$enable_alsa" = "x" || test "x$enable_alsa" = "xdefault" || test "x$enable_alsa" = "xyes"; then
diff --git a/data/themes/Makefile.am b/data/themes/Makefile.am
index a6e6cc5..854d2f0 100644
--- a/data/themes/Makefile.am
+++ b/data/themes/Makefile.am
@@ -1,7 +1,12 @@
 MAINTAINERCLEANFILES = Makefile.in
 SUBDIRS = images
 
+if CROSS_COMPILING
+EDJE_CC = @HOST_PROG_EDJE_CC@
+else
 EDJE_CC = @edje_cc@
+endif
+
 EDJE_FLAGS = \
 -v \
 -id $(top_srcdir)/data/themes/images \
