diff --git a/Makefile b/Makefile
index 0426e44..4d5cf6b 100644
--- a/Makefile
+++ b/Makefile
@@ -1,8 +1,14 @@
+#CROSS_COMPILE=armv7a-softfloat-linux-gnueabi-
+#KV_DIR=/usr/armv7a-softfloat-linux-gnueabi/usr/src/linux
 
-CC=$(CROSS_COMPILE)gcc
-CFLAGS=-I/home/valkeine/work/linux/usr/include -O2 -Wall
+CC=$(CHOST)-gcc
+LD=$(CHOST)-ld
 
-all: pan readback upd perf rect
+CFLAGS=-I/usr/armv7a-softfloat-linux-gnueabi/usr/src/linux/usr/include -O2 -Wall
 
+all: perf readback upd rect pan
+
+%.o: %.c
+	$(CC) $(CFLAGS) $(TARGET_ARCH) -c -o $@ $<
 clean:
 	rm -f pan readback upd perf rect
