--- /usr/portage/net-fs/nfs-utils/nfs-utils-1.2.0.ebuild	2009-06-02 21:03:30.000000000 +0200
+++ nfs-utils-1.2.0.ebuild	2009-08-09 23:00:39.000000000 +0200
@@ -4,7 +4,7 @@
 
 EAPI="1"
 
-inherit eutils flag-o-matic multilib
+inherit eutils flag-o-matic multilib toolchain-funcs
 
 DESCRIPTION="NFS client and server daemons"
 HOMEPAGE="http://linux-nfs.org/"
@@ -54,7 +54,12 @@
 		$(use_enable nfsv4) \
 		$(use_enable ipv6) \
 		$(use nfsv4 && use_enable kerberos gss || echo "--disable-gss")
-	emake || die "Failed to compile"
+
+	if tc-is-cross-compiler; then
+		emake CC=$(tc-getCC) || die "Failed to compile"
+	else
+		emake || die "Failed to compile"
+	fi
 }
 
 src_install() {
